### ISSUE TITLE
Your results versus ours

### ISSUE THREAD

**Issue by StrangeTcy** [epfml/dynamic-sparse-flash-attention#2]
We've run the experiments from your runtime experiments folder, and got the results plotted here:

[fwd+bwd-hash-sparse-vs-full-flash.pdf](https://github.com/epfml/dynamic-sparse-flash-attention/files/11799097/fwd%2Bbwd-hash-sparse-vs-full-flash.pdf)
[fwd+bwd-qk-sparse-vs-full-flash.pdf](https://github.com/epfml/dynamic-sparse-flash-attention/files/11799098/fwd%2Bbwd-qk-sparse-vs-full-flash.pdf)

The short version: we have flash attention results pretty close to hash-sparse, and you do not (according to the plots in the notebook). Is there a clear reason for that?
Did pytorch's `torch.nn.functional.scaled_dot_product_attention` get better over time?

I can post the versions of the libraries used and provide any other details you may need.

**Comment by mpagli** [comment]
To sum up the comparison:

### For hash-sparse:

* from the paper:
<img width="543" alt="image" src="https://github.com/epfml/dynamic-sparse-flash-attention/assets/2898358/11a7c20b-65ae-4519-953e-77c9ecc28f87">

* Yours:
<img width="680" alt="image" src="https://github.com/epfml/dynamic-sparse-flash-attention/assets/2898358/3c183c3d-98f2-4fe7-95a7-09182601dd7e">

### For QK-sparse:

* from the paper:
<img width="379" alt="image" src="https://github.com/epfml/dynamic-sparse-flash-attention/assets/2898358/3b0e4d8f-1964-4e25-9573-b48961b4117f">

* Yours:
<img width="445" alt="image" src="https://github.com/epfml/dynamic-sparse-flash-attention/assets/2898358/f6eac90e-0f21-41f3-bb11-e67bb566d568">

It seems to me `torch.nn.functional.scaled_dot_product_attention` got slower over time, no? That is nonetheless very weird. What is your setup?

**Comment by StrangeTcy** [comment]
"It seems to me torch.nn.functional.scaled_dot_product_attention got slower over time, no?"
-- probably, but since I haven't followed its development that closely for any considerable length of time, I wouldn't know why that might have happened.

My setup is as follows:
* I've cloned the repo and created a new conda env for it, and in it I just `pip installed` all the requirements
* the system being used is Ubuntu 20.04 and the GPU used to run these tests was an Nvidia A100 (`NVIDIA A100 80GB PCIe` as reported by `nvidia-smi`), with CUDA ver. 12.1

**Comment by mpagli** [comment]
We used CUDA 11.7 and an A100 40GB. Those are the only differences I see. I would try changing CUDA version. Or using a more recent pytorch version maybe more optimized for cuda 12.1?

**Comment by StrangeTcy** [comment]
Very well, I might reopen the issue once more data becomes available.
